sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcSchemaIOProvider.java

```diff
@@ -0,0 +1,172 @@
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.Schema.FieldType;
+import org.apache.beam.sdk.schemas.io.SchemaIO;
+import org.apache.beam.sdk.schemas.io.SchemaIOProvider;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.beam.sdk.values.Row;
+
+/**
+ * An implementation of {@link SchemaIOProvider} for reading and writing JSON payloads with {@link
+ * JdbcIO}.
+ */
+@Internal
+@AutoService(SchemaIOProvider.class)
+public class JdbcSchemaIOProvider implements SchemaIOProvider {
+
+  /** Returns an id that uniquely represents this IO. */
+  @Override
+  public String identifier() {
+    return "jdbc";
+  }
+
+  /**
+   * Returns the expected schema of the configuration object. Note this is distinct from the schema
+   * of the data source itself.
+   */
+  @Override
+  public Schema configurationSchema() {
+    return Schema.builder()
+        .addStringField("driverClassName")
+        .addStringField("jdbcUrl")
+        .addStringField("username")
+        .addStringField("password")
+        .addNullableField("connectionProperties", FieldType.STRING)
+        .addNullableField("connectionInitSqls", FieldType.iterable(FieldType.STRING))
+        .addNullableField("readQuery", FieldType.STRING)
+        .addNullableField("writeStatement", FieldType.STRING)
+        .addNullableField("fetchSize", FieldType.INT16)
+        .addNullableField("outputParallelization", FieldType.BOOLEAN)
+        .build();
+  }
+
+  /**
+   * Produce a SchemaIO given a String representing the data's location, the schema of the data that
+   * resides there, and some IO-specific configuration object.
+   */
+  @Override
+  public JdbcSchemaIO from(String location, Row configuration, Schema dataSchema) {
+    return new JdbcSchemaIO(location, configuration);
+  }
+
+  @Override
+  public boolean requiresDataSchema() {
+    return false;
+  }
+
+  @Override
+  public PCollection.IsBounded isBounded() {
+    return PCollection.IsBounded.BOUNDED;
+  }
+
+  /** An abstraction to create schema aware IOs. */
+  static class JdbcSchemaIO implements SchemaIO, Serializable {
+    protected final Row config;
+    protected final String location;
+
+    JdbcSchemaIO(String location, Row config) {
+      this.config = config;
+      this.location = location;
+    }
+
+    @Override
+    public Schema schema() {
+      return null;
+    }
+
+    @Override
+    public PTransform<PBegin, PCollection<Row>> buildReader() {
+      String readQuery;
+      if (config.getString("readQuery") != null) {
+        readQuery = config.getString("readQuery");
+      } else {
+        readQuery = String.format("SELECT f_int FROM %s", location);
```

Comment:
Similarly this shouldn't mention specific field names. In this case I think we could just do `SELECT * FROM %s`

Reply:
Done.